relengapi/blueprints/slaveloan/__init__.py

```diff
@@ -0,0 +1,233 @@
+from relengapi.blueprints.slaveloan.model import Loans
+from relengapi.blueprints.slaveloan.model import Machines
+
+logger = logging.getLogger(__name__)
+
+bp = Blueprint('slaveloan', __name__,
+               template_folder='templates',
+               static_folder='static')
+
+p.slaveloan.admin.doc("Administer Slaveloans for all users")
+
+##################
+#  RESTful APIs  #
+##################
+
+
+@bp.route('/loans/')
+@p.slaveloan.admin.require()
+@apimethod([rest.Loan])
+def get_loans():
+    "Get the list of all `active` loans you can see"
+    # XXX: Use permissions to filter if not an admin
+    session = g.db.session('relengapi')
+    loans = session.query(Loans).filter(Loans.machine_id.isnot(None))
+    return [l.to_wsme() for l in loans.all()]
+
+
+@bp.route('/loans/<int:loanid>')
+@p.slaveloan.admin.require()
+@apimethod(rest.Loan, int)
+def get_loan(loanid):
+    "Get the details of a loan, by id"
+    # XXX: Use permissions to ensure admin | loanee
+    session = g.db.session('relengapi')
+    l = session.query(Loans).get(loanid)
+    return l.to_wsme()
+
+
+@bp.route('/loans/<int:loanid>/history')
+@p.slaveloan.admin.require()
+@apimethod([rest.HistoryEntry], int)
+def get_loan_history(loanid):
+    "Get the history associated with this loan"
+    # XXX: Use permissions to ensure admin | loanee
+    session = g.db.session('relengapi')
+    histories = session.query(History) \
+                       .filter(History.loan_id == loanid) \
+                       .order_by(asc(History.timestamp))
+    return [h.to_wsme() for h in histories.all()]
+
+
+@bp.route('/loans/all')
+@p.slaveloan.admin.require()
+@apimethod([rest.Loan])
+def get_all_loans():
+    "Get the list of all loans you can see"
+    # XXX: Use permissions to filter if not an admin
+    session = g.db.session('relengapi')
+    loans = session.query(Loans)
+    return [l.to_wsme() for l in loans.all()]
+
+
+@bp.route('/loans/new', methods=['POST'])
+@p.slaveloan.admin.require()
+@apimethod(None, body=rest.LoanAdminRequest)
+def new_loan_from_admin(body):
+    "Creates a new loan entry"
+    if not body.status:
+        raise BadRequest("Missing Status Field")
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.bugzilla_email:
+        raise BadRequest("Missing Bugzilla E-Mail")
+    if body.status != 'PENDING':
+        if not body.fqdn:
+            raise BadRequest("Missing Machine FQDN")
+        if not body.ipaddr:
+            raise BadRequest("Missing Machine IP Address")
+
+    session = g.db.session('relengapi')
+    try:
+        if body.status != 'PENDING':
+            m = Machines.as_unique(session, fqdn=body.fqdn,
+                                   ipaddr=body.ipaddr)
+        h = Humans.as_unique(session, ldap=body.LDAP,
+                             bugzilla=body.bugzilla)
+    except sa.exc.IntegrityError:
+        raise InternalServerError("Integrity Error from Database, please retry.")
```

Comment:
This is eating the IntegrityError and replacing it with something much less useful to the user.  Better to just let the IntegrityError bubble up.